File: trevrpc-go/cmd/trevrpc-conformance-go/protocol.go
package main

import (
	"encoding/hex"
	"errors"
	"fmt"
	"strconv"
	"strings"

	trevrpc "trev.zip/llc/trevrpc/trevrpc-go"
)

const protocolVersion = 1

type metadataEntry struct {
	KeyHex   string `json:"key_hex"`
	ValueHex string `json:"value_hex"`
}

type normalizedRequest struct {
	Type         string          `json:"type"`
	ServiceHex   string          `json:"service_hex"`
	MethodHex    string          `json:"method_hex"`
	BodyHex      string          `json:"body_hex"`
	Metadata     []metadataEntry `json:"metadata"`
	Kind         string          `json:"kind"`
	Version      string          `json:"version"`
	TimeoutNanos string          `json:"timeout_nanos"`
}

type normalizedResponse struct {
	Type       string          `json:"type"`
	StatusRaw  string          `json:"status_raw"`
	StatusCode uint32          `json:"status_code"`
	MessageHex string          `json:"message_hex"`
	BodyHex    string          `json:"body_hex"`
	Metadata   []metadataEntry `json:"metadata"`
}

type normalizedStreamFrame struct {
	Type       string          `json:"type"`
	Kind       string          `json:"kind"`
	KindRaw    string          `json:"kind_raw"`
	StatusRaw  string          `json:"status_raw"`
	StatusCode uint32          `json:"status_code"`
	MessageHex string          `json:"message_hex"`
	BodyHex    string          `json:"body_hex"`
	Metadata   []metadataEntry `json:"metadata"`
}

type runCommand struct {
	Sequence     string
	CaseID       string
	Operation    string
	MessageType  string
	Message      any
	Body         []byte
	MaxFrameSize int
	Chunks       [][]byte
	Frames       [][]byte
}

type fieldParser struct {
	fields []string
	next   int
}

func parseCommand(line string) (*runCommand, bool, error) {
	if line == "STOP" {
		return nil, true, nil
	}
	if strings.ContainsRune(line, '\r') || !isASCII(line) {
		return nil, false, errors.New("command must be LF-terminated tab-delimited ASCII")
	}
	fields := strings.Split(line, "\t")
	if len(fields) < 4 || fields[0] != "RUN" {
		return nil, false, errors.New("expected RUN command")
	}
	if _, err := parseDecimal(fields[1], 64); err != nil {
		return nil, false, fmt.Errorf("invalid sequence: %w", err)
	}
	if !validID(fields[2]) {
		return nil, false, errors.New("invalid case ID")
	}
	parser := &fieldParser{fields: fields, next: 4}
	command := &runCommand{Sequence: fields[1], CaseID: fields[2], Operation: fields[3]}
	var err error
	switch command.Operation {
	case "codec.encode":
		command.MessageType, err = parser.token()
		if err == nil {
			command.Message, err = parser.message(command.MessageType)
		}
	case "codec.decode":
		command.MessageType, err = parser.messageType()
		if err == nil {
			command.Body, err = parser.hexBytes()
		}
	case "framing.encode":
		command.MessageType, err = parser.messageType()
		if err == nil {
			command.MaxFrameSize, err = parser.nonnegativeInt()
		}
		if err == nil {
			command.Message, err = parser.message(command.MessageType)
		}
	case "framing.decode_stream":
		command.MessageType, err = parser.messageType()
		if err == nil {
			command.MaxFrameSize, err = parser.nonnegativeInt()
		}
		if err == nil {
			command.Chunks, err = parser.hexList()
		}
	case "state.server_stream", "state.client_stream":
		command.Frames, err = parser.hexList()
	default:
		return nil, false, fmt.Errorf("unknown operation %q", command.Operation)
	}
	if err != nil {
		return nil, false, err
	}
	if parser.next != len(parser.fields) {
		return nil, false, errors.New("unexpected command fields")
	}
	return command, false, nil
}

func (p *fieldParser) token() (string, error) {
	if p.next >= len(p.fields) {
		return "", errors.New("missing command field")
	}
	value := p.fields[p.next]
	p.next++
	return value, nil
}

func (p *fieldParser) messageType() (string, error) {
	value, err := p.token()
	if err != nil {
		return "", err
	}
	if value != "rpc_request" && value != "rpc_response" && value != "rpc_stream_frame" {
		return "", fmt.Errorf("unknown message type %q", value)
	}
	return value, nil
}

func (p *fieldParser) hexBytes() ([]byte, error) {
	value, err := p.token()
	if err != nil {
		return nil, err
	}
	return decodeLowerHex(value)
}

func (p *fieldParser) nonnegativeInt() (int, error) {
	value, err := p.token()
	if err != nil {
		return 0, err
	}
	parsed, err := parseDecimal(value, strconv.IntSize)
	if err != nil {
		return 0, fmt.Errorf("invalid integer: %w", err)
	}
	return int(parsed), nil
}

func (p *fieldParser) remaining() int { return len(p.fields) - p.next }

func (p *fieldParser) hexList() ([][]byte, error) {
	count, err := p.nonnegativeInt()
	if err != nil {
		return nil, err
	}
	if count > p.remaining() {
		return nil, errors.New("list count exceeded remaining command fields")
	}
	values := make([][]byte, count)
	for index := range values {
		values[index], err = p.hexBytes()
		if err != nil {
			return nil, err
		}
	}
	return values, nil
}

func (p *fieldParser) message(messageType string) (any, error) {
	switch messageType {
	case "rpc_request":
		service, err := p.hexBytes()
		if err != nil {
			return nil, err
		}
		method, err := p.hexBytes()
		if err != nil {
			return nil, err
		}
		body, err := p.hexBytes()
		if err != nil {
			return nil, err
		}
		metadata, entries, err := p.metadata()
		if err != nil {
			return nil, err
		}
		kind, err := p.token()
		if err != nil {
			return nil, err
		}
		kindValue, ok := map[string]trevrpc.RpcKind{"unary": trevrpc.RpcKindUnary, "client_stream": trevrpc.RpcKindClientStreaming, "server_stream": trevrpc.RpcKindServerStreaming, "bidi": trevrpc.RpcKindBidirectionalStreaming}[kind]
		if !ok {
			return nil, fmt.Errorf("unknown RPC kind %q", kind)
		}
		versionText, err := p.token()
		if err != nil {
			return nil, err
		}
		version, err := parseDecimal(versionText, 32)
		if err != nil {
			return nil, err
		}
		timeoutText, err := p.token()
		if err != nil {
			return nil, err
		}
		timeout, err := parseDecimal(timeoutText, 64)
		if err != nil {
			return nil, err
		}
		return &trevrpc.RpcRequest{Service: string(service), Method: string(method), Body: body, Metadata: metadata, Kind: kindValue, Version: uint32(version), TimeoutNanos: timeout}, validateMetadataOrder(entries)
	case "rpc_response":
		statusText, err := p.token()
		if err != nil {
			return nil, err
		}
		status, err := parseDecimal(statusText, 32)
		if err != nil {
			return nil, err
		}
		message, err := p.hexBytes()
		if err != nil {
			return nil, err
		}
		body, err := p.hexBytes()
		if err != nil {
			return nil, err
		}
		metadata, entries, err := p.metadata()
		if err != nil {
			return nil, err
		}
		return &trevrpc.RpcResponse{Status: uint32(status), Message: string(message), Body: body, Metadata: metadata}, validateMetadataOrder(entries)
	case "rpc_stream_frame":
		kindText, err := p.token()
		if err != nil {
			return nil, err
		}
		kind, err := parseDecimal(kindText, 32)
		if err != nil {
			return nil, err
		}
		statusText, err := p.token()
		if err != nil {
			return nil, err
		}
		status, err := parseDecimal(statusText, 32)
		if err != nil {
			return nil, err
		}
		message, err := p.hexBytes()
		if err != nil {
			return nil, err
		}
		body, err := p.hexBytes()
		if err != nil {
			return nil, err
		}
		metadata, entries, err := p.metadata()
		if err != nil {
			return nil, err
		}
		return &trevrpc.RpcStreamFrame{Kind: trevrpc.RpcStreamFrameKind(kind), Status: uint32(status), Message: string(message), Body: body, Metadata: metadata}, validateMetadataOrder(entries)
	default:
		return nil, fmt.Errorf("unknown message type %q", messageType)
	}
}

func (p *fieldParser) metadata() (trevrpc.Metadata, []metadataEntry, error) {
	count, err := p.nonnegativeInt()
	if err != nil {
		return nil, nil, err
	}
	if count > p.remaining()/2 {
		return nil, nil, errors.New("metadata count exceeded remaining command fields")
	}
	metadata := trevrpc.Metadata{}
	entries := make([]metadataEntry, 0, count)
	for range count {
		key, err := p.hexBytes()
		if err != nil {
			return nil, nil, err
		}
		value, err := p.hexBytes()
		if err != nil {
			return nil, nil, err
		}
		entry := metadataEntry{KeyHex: hex.EncodeToString(key), ValueHex: hex.EncodeToString(value)}
		entries = append(entries, entry)
		metadata[string(key)] = value
	}
	return metadata, entries, nil
}

func parseDecimal(value string, bits int) (uint64, error) {
	if value == "" || (len(value) > 1 && value[0] == '0') {
		return 0, errors.New("non-canonical decimal")
	}
	for index := range len(value) {
		if value[index] < '0' || value[index] > '9' {
			return 0, errors.New("non-decimal value")
		}
	}
	return strconv.ParseUint(value, 10, bits)
}

func decodeLowerHex(value string) ([]byte, error) {
	if len(value)%2 != 0 {
		return nil, errors.New("hex value has odd length")
	}
	for index := range len(value) {
		if !((value[index] >= '0' && value[index] <= '9') || (value[index] >= 'a' && value[index] <= 'f')) {
			return nil, errors.New("hex value must be lowercase")
		}
	}
	return hex.DecodeString(value)
}

func validID(value string) bool {
	if value == "" {
		return false
	}
	for index := range len(value) {
		byteValue := value[index]
		if !((byteValue >= 'a' && byteValue <= 'z') || (byteValue >= '0' && byteValue <= '9') || byteValue == '.' || byteValue == '_' || byteValue == '-') {
			return false
		}
	}
	return true
}

func isASCII(value string) bool {
	for index := range len(value) {
		if value[index] > 0x7f {
			return false
		}
	}
	return true
}

func validateMetadataOrder(entries []metadataEntry) error {
	for index := 1; index < len(entries); index++ {
		if entries[index-1].KeyHex >= entries[index].KeyHex {
			return errors.New("metadata entries must be key-sorted and unique")
		}
	}
	return nil
}

File: trevrpc-kotlin/conformance-peer/src/main/kotlin/zip/trev/trevrpc/conformance/Protocol.kt
package zip.trev.trevrpc.conformance

import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction

internal const val PROTOCOL_VERSION = 1
internal const val MAX_COMMAND_BYTES = 262_144
internal const val MAX_EVENT_BYTES = 65_536
internal const val PEER_ID = "kotlin"

internal val capabilities =
    listOf(
        "codec.decode",
        "codec.encode",
        "framing.decode_stream",
        "framing.encode",
        "state.client_stream",
        "state.server_stream",
    )

internal data class MetadataEntry(
    val key: ByteArray,
    val value: ByteArray,
)

internal sealed interface NormalizedMessage {
    data class Request(
        val service: ByteArray,
        val method: ByteArray,
        val body: ByteArray,
        val metadata: List<MetadataEntry>,
        val kind: String,
        val version: UInt,
        val timeoutNanos: ULong,
    ) : NormalizedMessage

    data class Response(
        val statusRaw: UInt,
        val message: ByteArray,
        val body: ByteArray,
        val metadata: List<MetadataEntry>,
    ) : NormalizedMessage

    data class StreamFrame(
        val kindRaw: UInt,
        val statusRaw: UInt,
        val message: ByteArray,
        val body: ByteArray,
        val metadata: List<MetadataEntry>,
    ) : NormalizedMessage
}

internal data class RunCommand(
    val sequence: String,
    val caseId: String,
    val operation: String,
    val messageType: String? = null,
    val message: NormalizedMessage? = null,
    val body: ByteArray? = null,
    val maxFrameSize: Int? = null,
    val chunks: List<ByteArray> = emptyList(),
    val frames: List<ByteArray> = emptyList(),
)

internal sealed interface ParsedCommand {
    data object Stop : ParsedCommand

    data class Run(
        val command: RunCommand,
    ) : ParsedCommand
}

internal fun parseCommand(line: ByteArray): ParsedCommand {
    if (line.any { byte -> byte.toInt() < 0 || byte.toInt() == '\r'.code }) {
        protocolFailure("command must be LF-terminated tab-delimited ASCII")
    }
    val text = line.toString(Charsets.US_ASCII)
    if (text == "STOP") return ParsedCommand.Stop
    val fields = text.split('\t')
    if (fields.size < 4 || fields[0] != "RUN") protocolFailure("expected RUN command")
    parseUnsigned(fields[1], ULong.MAX_VALUE, "sequence")
    if (!validId(fields[2])) protocolFailure("invalid case ID")

    val parser = FieldParser(fields, 4)
    val base = RunCommand(sequence = fields[1], caseId = fields[2], operation = fields[3])
    val command =
        when (base.operation) {
            "codec.encode" -> {
                val messageType = parser.messageType()
                base.copy(messageType = messageType, message = parser.message(messageType))
            }

            "codec.decode" -> {
                base.copy(
                    messageType = parser.messageType(),
                    body = parser.hexBytes(),
                )
            }

            "framing.encode" -> {
                val messageType = parser.messageType()
                base.copy(
                    messageType = messageType,
                    maxFrameSize = parser.nonnegativeInt(),
                    message = parser.message(messageType),
                )
            }

            "framing.decode_stream" -> {
                base.copy(
                    messageType = parser.messageType(),
                    maxFrameSize = parser.nonnegativeInt(),
                    chunks = parser.hexList(),
                )
            }

            "state.server_stream", "state.client_stream" -> {
                base.copy(frames = parser.hexList())
            }

            else -> {
                protocolFailure("unknown operation ${base.operation}")
            }
        }
    if (!parser.exhausted) protocolFailure("unexpected command fields")
    return ParsedCommand.Run(command)
}

internal class StrictLineReader(
    private val input: InputStream,
    private val maximum: Int = MAX_COMMAND_BYTES,
) {
    fun readLine(): ByteArray {
        val output = ByteArrayOutputStream()
        while (true) {
            val value = input.read()
            if (value == -1) {
                if (output.size() == 0) protocolFailure("controller input ended without STOP")
                protocolFailure("command was not LF-terminated")
            }
            if (value == '\n'.code) return output.toByteArray()
            if (output.size() == maximum) protocolFailure("command line exceeded limit")
            output.write(value)
        }
    }
}

internal fun ByteArray.strictUtf8(fieldName: String): String =
    try {
        Charsets.UTF_8
            .newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(this))
            .toString()
    } catch (error: Exception) {
        throw IllegalArgumentException("$fieldName was not valid UTF-8", error)
    }

internal fun ByteArray.lowerHex(): String {
    val alphabet = "0123456789abcdef"
    return CharArray(size * 2)
        .also { encoded ->
            forEachIndexed { index, byte ->
                val value = byte.toInt() and 0xff
                encoded[index * 2] = alphabet[value ushr 4]
                encoded[index * 2 + 1] = alphabet[value and 0x0f]
            }
        }.concatToString()
}

private class FieldParser(
    private val fields: List<String>,
    private var next: Int,
) {
    val exhausted: Boolean
        get() = next == fields.size

    private val remaining: Int
        get() = fields.size - next

    fun token(): String {
        if (next >= fields.size) protocolFailure("missing command field")
        return fields[next++]
    }

    fun messageType(): String {
        val value = token()
        if (value !in setOf("rpc_request", "rpc_response", "rpc_stream_frame")) {
            protocolFailure("unknown message type $value")
        }
        return value
    }

    fun hexBytes(): ByteArray = decodeLowerHex(token())

    fun nonnegativeInt(): Int = parseUnsigned(token(), Int.MAX_VALUE.toULong(), "integer").toInt()

    fun hexList(): List<ByteArray> {
        val count = nonnegativeInt()
        if (count > remaining) protocolFailure("list count exceeded remaining command fields")
        return buildList(count) {
            repeat(count) { add(hexBytes()) }
        }
    }

    fun message(messageType: String): NormalizedMessage =
        when (messageType) {
            "rpc_request" -> {
                NormalizedMessage.Request(
                    service = hexBytes(),
                    method = hexBytes(),
                    body = hexBytes(),
                    metadata = metadata(),
                    kind = rpcKind(),
                    version = uint32(),
                    timeoutNanos = uint64(),
                )
            }

            "rpc_response" -> {
                NormalizedMessage.Response(
                    statusRaw = uint32(),
                    message = hexBytes(),
                    body = hexBytes(),
                    metadata = metadata(),
                )
            }

            "rpc_stream_frame" -> {
                NormalizedMessage.StreamFrame(
                    kindRaw = uint32(),
                    statusRaw = uint32(),
                    message = hexBytes(),
                    body = hexBytes(),
                    metadata = metadata(),
                )
            }

            else -> {
                protocolFailure("unknown message type $messageType")
            }
        }

    private fun metadata(): List<MetadataEntry> {
        val count = nonnegativeInt()
        if (count > remaining / 2) protocolFailure("metadata count exceeded remaining command fields")
        val entries =
            buildList(count) {
                repeat(count) { add(MetadataEntry(hexBytes(), hexBytes())) }
            }
        entries.zipWithNext().forEach { (left, right) ->
            if (compareBytes(left.key, right.key) >= 0) {
                protocolFailure("metadata entries must be key-sorted and unique")
            }
        }
        return entries
    }

    private fun rpcKind(): String {
        val value = token()
        if (value !in setOf("unary", "client_stream", "server_stream", "bidi")) {
            protocolFailure("unknown RPC kind $value")
        }
        return value
    }

    private fun uint32(): UInt = parseUnsigned(token(), UInt.MAX_VALUE.toULong(), "uint32").toUInt()

    private fun uint64(): ULong = parseUnsigned(token(), ULong.MAX_VALUE, "uint64")
}

private fun decodeLowerHex(value: String): ByteArray {
    if (value.length % 2 != 0) protocolFailure("hex value has odd length")
    if (value.any { it !in '0'..'9' && it !in 'a'..'f' }) protocolFailure("hex value must be lowercase")
    return ByteArray(value.length / 2) { index ->
        value.substring(index * 2, index * 2 + 2).toInt(16).toByte()
    }
}

private fun parseUnsigned(
    value: String,
    maximum: ULong,
    fieldName: String,
): ULong {
    if (value.isEmpty() || (value.length > 1 && value[0] == '0')) protocolFailure("non-canonical $fieldName decimal")
    if (value.any { it !in '0'..'9' }) protocolFailure("non-decimal $fieldName value")
    val parsed = value.toULongOrNull() ?: protocolFailure("$fieldName exceeded supported range")
    if (parsed > maximum) protocolFailure("$fieldName exceeded supported range")
    return parsed
}

private fun validId(value: String): Boolean =
    value.isNotEmpty() &&
        value.all { character ->
            character in 'a'..'z' || character in '0'..'9' || character in "._-"
        }

private fun compareBytes(
    left: ByteArray,
    right: ByteArray,
): Int {
    for (index in 0 until minOf(left.size, right.size)) {
        val comparison = (left[index].toInt() and 0xff).compareTo(right[index].toInt() and 0xff)
        if (comparison != 0) return comparison
    }
    return left.size.compareTo(right.size)
}

internal fun protocolFailure(message: String): Nothing = throw ProtocolException(message)

internal class ProtocolException(
    message: String,
) : Exception(message)
